### PR TITLE
Replace FlexSlider $.fadeIn() with vanilla .animate()

### LINKED ISSUE #41976: Don't use $.fadeIn() in FlexSlider for better jQuery slim build compatibility
### Describe the solution you'd like

**jquery.flexslider.js** uses `$.fadeIn()` in 1 place. It can be easily replicated with a Vanilla JS .animate(). This will get FlexSlider closer to being compatible with the jQuery slim build, which excludes the [ajax](https://api.jquery.com/category/ajax/) and [effects](https://api.jquery.com/category/effects/) modules.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by apenchev**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #41976 and includes a small refactor so that the two `.animate()` calls are now using the same keyframes format.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

To test the `.fadeIn()` replace:

1. Paste the following code in an empty `index.html` file and put the [jquery.flexslider.js](https://github.com/woocommerce/woocommerce/blob/c52717d9ba8364c2a0db75755946461c89828139/plugins/woocommerce/client/legacy/js/flexslider/jquery.flexslider.js) file from this PR in the same folder:

```html
<!doctype html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>#41979 Demo</title>
  <style>.slides { margin: 0; padding: 0; list-style: none; }</style>
</head>
<body>

<div id="slider" class="flexslider">
  <ul class="slides">
    <li><img src="https://woo.com/wp-content/themes/woo/images/flex/flexslider_a.jpg"></li>
    <!--<li><img src="https://woo.com/wp-content/themes/woo/images/flex/flexslider_b.jpg"></li>
    <li><img src="https://woo.com/wp-content/themes/woo/images/flex/flexslider_c.jpg"></li>
    <li><img src="https://woo.com/wp-content/themes/woo/images/flex/flexslider_d.jpg"></li>-->
  </ul>
</div>

<script src="https://code.jquery.com/jquery-3.7.1.min.js" integrity="sha256-/JqT3SQfawRcv/BIHPThkBvs0OEvtFFmqPF/lYI/Cxo=" crossorigin="anonymous"></script>
<script src="./jquery.flexslider.js"></script>
<script>
$(function() {
  $('#slider').flexslider({
    animation: 'slide',
    allowOneSlide: false,
  });
});
</script>

</body>
</html>
```
2. Open `index.html` with any browser and confirm that there's a fadeIn animation for the single image on the page.

To test the `.animate()` keyframes refactor:

1. Add this custom code to your theme's `functions.php`:
```
function carousel_options( $options ) {
	return array_merge( $options, array( 'useCSS' => false ) );
}
add_filter( 'woocommerce_single_product_carousel_options', 'carousel_options' )
```
2. Navigate to a Single Product Page containing multiple images.
3. Confirm that you are able to switch between the multiple images.

### CODE REVIEW

File: plugins/woocommerce/client/legacy/js/flexslider/jquery.flexslider.js [hunk 1]
```diff
@@ -1134,7 +1132,8 @@
             $slides = $this.find(selector);
 
       if ( ( $slides.length === 1 && options.allowOneSlide === false ) || $slides.length === 0 ) {
-          $slides.fadeIn(400);
+          var fadeIn = [{ opacity: 0 }, { opacity: 1 }];
```
Comment (barryhughes) [inline plugins/woocommerce/client/legacy/js/flexslider/jquery.flexslider.js:1135]: I'm not sure I really understand how to test this change.

If I have a product image, and one gallery image, then `$slides.length` will be 2. If I have just one gallery image *or* just a product image (and no other gallery images), execution does not seem to reach this point. Though I did see your note which I think relates to this—

> By default, pages with a single image don't initialize FlexSlider at all, which begs the question of whether we should straight up delete the .fadeIn() code in the future.

—I'm a little reluctant to include this part of the change without clearer testing instructions, or a clearer understanding of when/how this would actually be invoked.

Comment (apenchev) [reply plugins/woocommerce/client/legacy/js/flexslider/jquery.flexslider.js:1135]: Have you deleted `1===this.$images.length&&(this.flexslider_enabled=!1),` from `plugins/woocommerce/assets/js/frontend/single-product.min.js` before testing? It's needed for FlexSlider to initialize for single image product (by default WooCommerce choses not to load FlexSlider for single images, which is sensible).

I've added some screenshots in the testing instructions. I'm using the Twenty Twenty-Four theme for testing, and the default **non-blockyfied** product page template. I only have a single product image, the way it comes with the default WooCommerce plugin install, I haven't set any gallery images. Here is how the product page looks in my editor:

stepqqq

**Comment by barryhughes** [comment]
> Have you deleted 1===this.$images.length&&(this.flexslider_enabled=!1), from plugins/woocommerce/assets/js/frontend/single-product.min.js before testing?

Just a quick note first that we can add `define( 'SCRIPT_DEBUG', true );` to our `wp-config.php` file, and then we can reference and reason about the logic based on the unminified versions of the script. In this case, for clarity, I think we are talking about [this block](https://github.com/woocommerce/woocommerce/blob/8.4.0/plugins/woocommerce/client/legacy/js/flexslider/jquery.flexslider.js#L1129-L1134):

```js

      if ( ( $slides.length === 1 && options.allowOneSlide === false ) || $slides.length === 0 ) {
          $slides.fadeIn(400);
          if (options.start) { options.start($this); }
        } else if ($this.data('flexslider') === undefined) {
          new $.flexslider(this, options);
        }
```

In any case: no, I did not delete that condition. My concern is that the logic [here](https://github.com/woocommerce/woocommerce/blob/8.4.0/plugins/woocommerce/client/legacy/js/frontend/single-product.js#L94-L98) and [here](https://github.com/woocommerce/woocommerce/blob/8.4.0/plugins/woocommerce/client/legacy/js/frontend/single-product.js#L115-L118) means the above code will not be executed at all (when there are not enough images on the page to require it), and so the suggested testing instructions seem too contrived for me to be confident in this part of the change.

Basically: the first half of the change that we currently see on [lines 711-714](https://github.com/woocommerce/woocommerce/blob/c52717d9ba8364c2a0db75755946461c89828139/plugins/woocommerce/client/legacy/js/flexslider/jquery.flexslider.js#L711-L714) seems fine, but I'm less confident of the second half (because we don't currently have a clear way to test this).

**Comment by barryhughes** [comment]
... I'm happy to be corrected. I'm just reluctant because if there is no way to test without directly changing the source as you described, then it opens up the possibility we are misunderstanding how it works and are going to cause disruption for someone out there (small and seemingly safe changes like this can sometimes 'backfire').

File: plugins/woocommerce/client/legacy/js/flexslider/jquery.flexslider.js (hunk 1 above)
Comment (barryhughes) [reply plugins/woocommerce/client/legacy/js/flexslider/jquery.flexslider.js:1135]: Sorry, accidentally replied [outside of this thread](https://github.com/woocommerce/woocommerce/pull/41979#issuecomment-1863029939).

**Comment by apenchev** [comment]
> ... I'm happy to be corrected. I'm just reluctant because if there is no way to test without directly changing the source as you described, then it opens up the possibility we are misunderstanding how it works and are going to cause disruption for someone out there (small and seemingly safe changes like this can sometimes 'backfire').

Fair enough. I don't believe there is a WooCommerce setup that can trigger the fadeIn code without us modifying `single-product.js`. I could create a custom HTML gallery example that triggers the fadeIn? If that is not satisfactory, I'll remove the fadeIn bit from this PR.

**Comment by barryhughes** [comment]
> I could create a custom HTML gallery example that triggers the fadeIn?

That sounds good to me :+1:

**Comment by apenchev** [comment]
I've edited the PR description to include an HTML example that triggers the fadeIn code.

File: plugins/woocommerce/changelog/flexslider-fadein [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: update
+
+FlexSlider $.fadeIn() is replaced with a Vanilla JS .animate().
```
Comment (barryhughes) [inline plugins/woocommerce/changelog/flexslider-fadein]:
```suggestion
Improved our FlexSlider implementation by removing the need for jQuery, in relation to fade-in animations.
```

**Review by barryhughes — APPROVED** [review]
Thank you, @apenchev. Waiting on automated checks completing, then (assuming there are no problems) we can go ahead with a merge.
